### PR TITLE
Add an ability to pass a function for severity secondary option

### PR DESCRIPTION
**Pull request by kizu**
Let me know if I should add more tests (if so, where?), or change something in the docs.

I did implement the feature as described in the https://github.com/stylelint/stylelint/issues/6863#issuecomment-1567238888 comment.

> Which issue, if any, is this issue related to?

Closes https://github.com/stylelint/stylelint/issues/6863

> Is there anything in the PR that needs further explanation?

No, it's self-explanatory.

### CODE REVIEW

File: types/stylelint/index.d.ts [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import type * as PostCSS from 'postcss';
 import type { GlobbyOptions } from 'globby';
 import type { cosmiconfig, TransformSync as CosmiconfigTransformSync } from 'cosmiconfig';
+import { Severity } from '../../lib/formatters/preprocessWarnings';
```
Comment (Mouvedia) [inline types/stylelint/index.d.ts:4]: Can you rename the import?
It's kinda confusing to have it twice in the same file.

see l72

**Comment by ybiquitous** [comment]
@kizu Thanks for creating this PR! I'm so afraid, but could you recreate this PR on the [`v16`](https://github.com/stylelint/stylelint/tree/v16) branch? Since we have already pushed many commits into the `v16` branch, we had to address conflicts if a big change would be merged into the `main` branch. 🙏🏼

File: types/stylelint/index.d.ts (hunk 1 above)
Comment (kizu) [reply types/stylelint/index.d.ts]: Fixed. I used the `stylelint.Severity` here instead of importing it from elsewhere.

**Comment by kizu** [comment]
@ybiquitous Sure, rebased to v16. Is there an ETA for when the v16 is expected to be released?

**Comment by ybiquitous** [comment]
@kizu Thank you. I'm sorry, but no ETA for now. If possible, I hope to release v16 this year, though.

File: .changeset/functional-severity.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+"stylelint": minor
+---
+
+Added: `severity` secondary option now accepts a function allowing to adjust the severity based on the rule's `messageArgs`.
```
Comment (ybiquitous) [inline .changeset/functional-severity.md:5]: [suggest] Could you simplify the description like this?

```suggestion
Added: `severity` secondary option's function support
```

**Review by ybiquitous — CHANGES_REQUESTED** [review]
No big problems, thank you so much.
I've left some minor suggestions, so I'd appreciate your addressing them. 🙏🏼

File: types/stylelint/index.d.ts [hunk 3]
```diff
@@ -203,6 +207,9 @@ declare namespace stylelint {
 		optional?: boolean;
 	};
 
+	/** @internal */
+	type RuleSeverity = Severity | RuleSeverityFunc;
```
Comment (Mouvedia) [inline types/stylelint/index.d.ts:211]: The type is hoisted I suppose but wouldn't it be clearer if it was set before line 145?

File: docs/user-guide/configure.md [hunk 4]
```diff
@@ -179,6 +179,41 @@ For example:
 
 Reporters may use these severity levels to display problems or exit the process differently.
 
+Experimental feature: some rules support [message arguments](#message). For these rules, it is possible to use a function for `severity`, which would accept these arguments, allowing you to adjust the severity based on these arguments.
```
Comment (Mouvedia) [inline docs/user-guide/configure.md:182]: suggestion

maybe a list would be better

**Review by ybiquitous — APPROVED** [review]
@kizu Thanks for addressing my reviews. LGTM! 👍🏼

File: docs/user-guide/configure.md (hunk 4 above)
Comment (Mouvedia) [reply docs/user-guide/configure.md:182]: It seemed to me that it emphasises too much that a rule without `message` cannot have `severity`.
Anyway it's minor; Ill resolve.

**Pull request merged**